taichi/analysis/same_statements.cpp

```diff
@@ -191,34 +216,68 @@ class IRNodeComparator : public IRVisitor {
     }
   }
 
-  static bool run(IRNode *root1, IRNode *root2) {
-    IRNodeComparator comparator(root2);
+  static bool run(IRNode *root1,
+                  IRNode *root2,
+                  const std::optional<std::unordered_map<int, int>> &id_map,
+                  bool check_same_value) {
+    IRNodeComparator comparator(root2, id_map, check_same_value);
     root1->accept(&comparator);
     return comparator.same;
   }
 };
 
 namespace irpass::analysis {
-bool same_statements(IRNode *root1, IRNode *root2) {
+bool same_statements(
+    IRNode *root1,
+    IRNode *root2,
+    const std::optional<std::unordered_map<int, int>> &id_map) {
+  // When id_map is std::nullopt by default, this function tests if
+  // root1 and root2 are the same, i.e., have the same type,
+  // the same operands and the same fields.
+  // If root1 and root2 are container statements or statement blocks,
+  // this function traverses the contents correspondingly.
+  // Two operands are considered the same if they have the same id
+  // and do not belong to either root, or they belong to root1 and root2
+  // according to the same position in the roots.
+  //
+  // For example, same_statements(block1, block2, std::nullopt) is true:
+  // <i32> $1 = ...
+  // block1 : {
+  //   <i32> $2 = const [1]
+  //   <i32> $3 = add $1 $2
+  // }
+  // block2 : {
+  //   <i32> $4 = const [1]
+  //   <i32> $5 = add $1 $4
+  // }
+  //
+  // If id_map is not std::nullopt, this function also recursively
+  // check the operands until ids in the id_map are reached.
+  // id_map is an id map from root1 to root2.
+  //
+  // In the above example, same_statements($3, $5, std::nullopt) is false
```

Comment:
Because $2 and $4 are considered different when `std::nullopt` is used as the third argument (the function will not check if they are the same recursively).

When comparing `block1` and `block2`, the function checks if $2 and $4 are the same.